### PR TITLE
V2 overhaul

### PR DESCRIPTION
**Pull request by robmarkcole**
Vastly simplify the main class and general improvements based on my experience with pandas, e.g. dropping use of multi-index, using long-form dataframe

### CODE REVIEW

File: detective/core.py [hunk 1]
```diff
@@ -72,320 +66,43 @@ def __init__(self, url, *, fetch_entities=True):
...
-        Basic query from list of entities. Must be from same domain.
-        Attempts to unpack lists up to 2 deep.
-
-        Parameters
-        ----------
-        entities : a list of entities
-
-        returns a df
-        """
+        entities = [e[0] for e in response]
```
Comment (dgomes) [inline detective/core.py]: can probably store entities directly into self.entities

File: detective/core.py [hunk 2]
```diff
@@ -72,320 +66,43 @@ def __init__(self, url, *, fetch_entities=True):
...
-            SELECT domain, entity_id, state, last_changed
+        query = f"""
+            SELECT domain, entity_id, state, last_changed, attributes
             FROM states
             WHERE
+                domain IN ('binary_sensor', 'sensor')
+            AND
                 state NOT IN ('unknown', 'unavailable')
+            AND
+                last_changed = last_updated
```
Comment (dgomes) [inline detective/core.py]: Sure about this ?

File: detective/functions.py [hunk 3]
```diff
@@ -1,24 +1,60 @@
...
-    except ValueError:
-        return False
+
+def format_binary_state(state: str):
+    """Return a binary for the state of binary sensors."""
+    if state == "on":
+        return 1
+    elif state == "off":
+        return 0
+    return state
```
Comment (dgomes) [inline detective/functions.py]: could use an exception if binary is not ON/OFF

File: detective/functions.py [hunk 4]
```diff
@@ -1,24 +1,60 @@
...
+
+
+def get_device_class(attributes: dict):
+    """Return the device class."""
+    device_class = attributes.get("device_class")
+    if device_class:
+        return device_class
+    return UNKNOWN
```
Comment (dgomes) [inline detective/functions.py]: return attributes.get("device_class", default=UNKNOWN)

File: detective/functions.py (hunk 3 above)
Comment (ties) [reply detective/functions.py]: Previously the binary version was returned as True/False, now as 0/1. I prefer the previous approach.

File: detective/core.py (hunk 2 above)
Comment (robmarkcole) [reply detective/core.py]: nope, its new so will remove

File: detective/functions.py (hunk 3 above)
Comment (robmarkcole) [reply detective/functions.py]: I am not really sure about this one. Encoding as 0/1 made sense to me as I am basically interested in aggregating on counts. Probably there is a better way using pandas, perhaps [get_dummies](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html) I will drop for now and give it some thought

File: detective/functions.py (hunk 4 above)
Comment (robmarkcole) [reply detective/functions.py]: yes!

File: detective/functions.py (hunk 3 above)
Comment (ties) [reply detective/functions.py]: Or plain group_by/count. `get_dummies` looks similar to the [OneHotEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) in scikit-learn.

Comment (robmarkcole) [reply detective/functions.py]: Encoding as 1/-1 would also make sense

**Pull request merged**